File: web/src/ipc/cursor_state.ts
// Shared hardware cursor descriptor layout.
//
// Keep this file in sync with:
//   crates/aero-shared/src/cursor_state.rs
//   emulator/protocol/aerogpu/aerogpu_pci.ts (AerogpuFormat)
//
// The state is stored in an `Int32Array` backed by a `SharedArrayBuffer` so it can be accessed
// from JS using `Atomics.*` operations.
//
// All fields are stored as 32-bit words. Most are logically `u32` but must be stored/read through
// `Int32Array` for `Atomics.wait` compatibility; therefore always reinterpret via `>>> 0`.
//
// Cursor position fields (`x`, `y`) are logically signed `i32` (allow negative coordinates) and
// should be read via `| 0`.

import { AerogpuFormat } from "../../../emulator/protocol/aerogpu/aerogpu_pci.ts";

// Cursor format values use the AeroGPU `AerogpuFormat` numeric (`u32`) discriminants.
//
// Semantics (from the AeroGPU protocol):
// - `*X8*` formats (`B8G8R8X8*`, `R8G8B8X8*`) do not carry alpha. When converting to RGBA
//   (e.g. for cursor blending), treat alpha as fully opaque (`0xff`) and ignore the stored
//   `X` byte.
// - `*_SRGB` variants are layout-identical to their UNORM counterparts; only the color space
//   interpretation differs. Presenters must avoid double-applying gamma when handling sRGB
//   cursor formats.
export const CURSOR_FORMAT_B8G8R8A8 = AerogpuFormat.B8G8R8A8Unorm;
export const CURSOR_FORMAT_B8G8R8X8 = AerogpuFormat.B8G8R8X8Unorm;
export const CURSOR_FORMAT_R8G8B8A8 = AerogpuFormat.R8G8B8A8Unorm;
export const CURSOR_FORMAT_R8G8B8X8 = AerogpuFormat.R8G8B8X8Unorm;

export const CURSOR_STATE_U32_LEN = 12 as const;
export const CURSOR_STATE_BYTE_LEN = CURSOR_STATE_U32_LEN * 4;

export const CURSOR_STATE_GENERATION_BUSY_BIT = 0x8000_0000 as const;

export const CursorStateIndex = {
  GENERATION: 0,
  ENABLE: 1,
  X: 2,
  Y: 3,
  HOT_X: 4,
  HOT_Y: 5,
  WIDTH: 6,
  HEIGHT: 7,
  PITCH_BYTES: 8,
  FORMAT: 9,
  BASE_PADDR_LO: 10,
  BASE_PADDR_HI: 11,
} as const;

export type CursorStateIndex = (typeof CursorStateIndex)[keyof typeof CursorStateIndex];

export interface CursorStateUpdate {
  enable: number;
  x: number;
  y: number;
  hotX: number;
  hotY: number;
  width: number;
  height: number;
  pitchBytes: number;
  format: number;
  basePaddrLo: number;
  basePaddrHi: number;
}

export interface CursorStateSnapshot extends CursorStateUpdate {
  generation: number;
}

export interface TrySnapshotCursorStateOptions {
  /**
   * Maximum number of seqlock read attempts before giving up.
   *
   * This is a hard bound; the function will never spin forever.
   */
  maxIterations?: number;
  /**
   * Optional wall-clock time budget (in milliseconds) before giving up.
   *
   * This bound is checked periodically (not every iteration) to avoid adding
   * significant overhead to successful snapshots.
   */
  maxMs?: number;
}

export function wrapCursorState(sab: SharedArrayBuffer, byteOffset = 0): Int32Array {
  if (!(sab instanceof SharedArrayBuffer)) {
    throw new TypeError("wrapCursorState requires a SharedArrayBuffer");
  }
  if (!Number.isFinite(byteOffset)) {
    throw new RangeError(`byteOffset must be a finite number, got ${String(byteOffset)}`);
  }
  const off = Math.trunc(byteOffset);
  if (off < 0) {
    throw new RangeError(`byteOffset must be >= 0, got ${off}`);
  }
  if (off % 4 !== 0) {
    throw new RangeError(`byteOffset must be 4-byte aligned, got ${off}`);
  }
  const requiredBytes = off + CURSOR_STATE_BYTE_LEN;
  if (requiredBytes > sab.byteLength) {
    throw new RangeError(
      `CursorState view would be out of bounds: need ${requiredBytes} bytes (offset=${off}, len=${CURSOR_STATE_BYTE_LEN}), sab.byteLength=${sab.byteLength}`,
    );
  }
  return new Int32Array(sab, off, CURSOR_STATE_U32_LEN);
}

export function cursorBasePaddr(snapshot: CursorStateSnapshot): bigint {
  return (BigInt(snapshot.basePaddrHi >>> 0) << 32n) | BigInt(snapshot.basePaddrLo >>> 0);
}

export function snapshotCursorState(words: Int32Array): CursorStateSnapshot {
  if (words.length < CURSOR_STATE_U32_LEN) {
    throw new RangeError(`CursorState Int32Array too small: len=${words.length}, need >=${CURSOR_STATE_U32_LEN}`);
  }

  // Seqlock-style snapshot using a busy bit.
  //
  // IMPORTANT: this must not spin forever if the writer crashes while holding the
  // busy bit. Bound the retry loop so callers (especially the GPU worker present
  // path) can recover and render a safe fallback.
  const startMs = typeof performance !== "undefined" && typeof performance.now === "function" ? performance.now() : Date.now();
  // Allow some time for a writer to complete even on slow/contended JS runtimes, but
  // still guarantee we won't spin forever if the busy bit is stuck.
  const MAX_SPIN_MS = 50;
  const MAX_SPINS = 1_000_000;

  for (let spins = 0; spins < MAX_SPINS; spins += 1) {
    const gen0 = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
    if ((gen0 & CURSOR_STATE_GENERATION_BUSY_BIT) !== 0) {
      if ((spins & 0x3fff) === 0) {
        const nowMs = typeof performance !== "undefined" && typeof performance.now === "function" ? performance.now() : Date.now();
        if (nowMs - startMs > MAX_SPIN_MS) break;
      }
      continue;
    }

    const enable = Atomics.load(words, CursorStateIndex.ENABLE) >>> 0;
    const x = Atomics.load(words, CursorStateIndex.X) | 0;
    const y = Atomics.load(words, CursorStateIndex.Y) | 0;
    const hotX = Atomics.load(words, CursorStateIndex.HOT_X) >>> 0;
    const hotY = Atomics.load(words, CursorStateIndex.HOT_Y) >>> 0;
    const width = Atomics.load(words, CursorStateIndex.WIDTH) >>> 0;
    const height = Atomics.load(words, CursorStateIndex.HEIGHT) >>> 0;
    const pitchBytes = Atomics.load(words, CursorStateIndex.PITCH_BYTES) >>> 0;
    const format = Atomics.load(words, CursorStateIndex.FORMAT) >>> 0;
    const basePaddrLo = Atomics.load(words, CursorStateIndex.BASE_PADDR_LO) >>> 0;
    const basePaddrHi = Atomics.load(words, CursorStateIndex.BASE_PADDR_HI) >>> 0;

    const gen1 = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
    if (gen0 !== gen1) {
      if ((spins & 0x3fff) === 0) {
        const nowMs = typeof performance !== "undefined" && typeof performance.now === "function" ? performance.now() : Date.now();
        if (nowMs - startMs > MAX_SPIN_MS) break;
      }
      continue;
    }

    return {
      generation: gen0,
      enable,
      x,
      y,
      hotX,
      hotY,
      width,
      height,
      pitchBytes,
      format,
      basePaddrLo,
      basePaddrHi,
    };
  }

  throw new Error("snapshotCursorState: timed out (writer busy bit stuck or update rate too high)");
}

function nowMs(): number {
  // `performance.now()` exists in browsers and modern Node. Fall back to `Date.now()`
  // so this helper is safe in minimal test environments.
  const perf = (globalThis as unknown as { performance?: { now?: () => number } }).performance;
  return typeof perf?.now === "function" ? perf.now() : Date.now();
}

/**
 * Attempt to snapshot cursor state without risking an infinite spin if the writer
 * wedge/crashes while holding the busy bit.
 *
 * Returns `null` when the snapshot could not be obtained within the configured bounds.
 */
export function trySnapshotCursorState(words: Int32Array, options?: TrySnapshotCursorStateOptions): CursorStateSnapshot | null {
  if (words.length < CURSOR_STATE_U32_LEN) {
    throw new RangeError(`CursorState Int32Array too small: len=${words.length}, need >=${CURSOR_STATE_U32_LEN}`);
  }

  const maxIterationsRaw = options?.maxIterations;
  const maxIterations =
    maxIterationsRaw === undefined
      ? 1000
      : (() => {
          if (!Number.isFinite(maxIterationsRaw)) {
            throw new RangeError(`trySnapshotCursorState maxIterations must be a finite number, got ${String(maxIterationsRaw)}`);
          }
          return Math.max(0, Math.trunc(maxIterationsRaw));
        })();

  const maxMsRaw = options?.maxMs;
  const deadlineMs =
    maxMsRaw === undefined
      ? Number.POSITIVE_INFINITY
      : (() => {
          if (!Number.isFinite(maxMsRaw)) {
            throw new RangeError(`trySnapshotCursorState maxMs must be a finite number, got ${String(maxMsRaw)}`);
          }
          if (maxMsRaw <= 0) return nowMs();
          return nowMs() + maxMsRaw;
        })();

  const CHECK_DEADLINE_EVERY = 64;

  // Seqlock-style snapshot using a busy bit, but with bounded retries.
  for (let it = 0; it < maxIterations; it += 1) {
    if (deadlineMs !== Number.POSITIVE_INFINITY && it % CHECK_DEADLINE_EVERY === 0) {
      if (nowMs() >= deadlineMs) return null;
    }

    const gen0 = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
    if ((gen0 & CURSOR_STATE_GENERATION_BUSY_BIT) !== 0) {
      continue;
    }

    const enable = Atomics.load(words, CursorStateIndex.ENABLE) >>> 0;
    const x = Atomics.load(words, CursorStateIndex.X) | 0;
    const y = Atomics.load(words, CursorStateIndex.Y) | 0;
    const hotX = Atomics.load(words, CursorStateIndex.HOT_X) >>> 0;
    const hotY = Atomics.load(words, CursorStateIndex.HOT_Y) >>> 0;
    const width = Atomics.load(words, CursorStateIndex.WIDTH) >>> 0;
    const height = Atomics.load(words, CursorStateIndex.HEIGHT) >>> 0;
    const pitchBytes = Atomics.load(words, CursorStateIndex.PITCH_BYTES) >>> 0;
    const format = Atomics.load(words, CursorStateIndex.FORMAT) >>> 0;
    const basePaddrLo = Atomics.load(words, CursorStateIndex.BASE_PADDR_LO) >>> 0;
    const basePaddrHi = Atomics.load(words, CursorStateIndex.BASE_PADDR_HI) >>> 0;

    const gen1 = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
    if (gen0 !== gen1) {
      continue;
    }

    return {
      generation: gen0,
      enable,
      x,
      y,
      hotX,
      hotY,
      width,
      height,
      pitchBytes,
      format,
      basePaddrLo,
      basePaddrHi,
    };
  }

  return null;
}

export function publishCursorState(words: Int32Array, update: CursorStateUpdate): number {
  if (words.length < CURSOR_STATE_U32_LEN) {
    throw new RangeError(`CursorState Int32Array too small: len=${words.length}, need >=${CURSOR_STATE_U32_LEN}`);
  }

  // Acquire the writer lock by setting the busy bit.
  let start = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
  while (true) {
    if ((start & CURSOR_STATE_GENERATION_BUSY_BIT) !== 0) {
      start = Atomics.load(words, CursorStateIndex.GENERATION) >>> 0;
      continue;
    }
    const desired = (start | CURSOR_STATE_GENERATION_BUSY_BIT) >>> 0;
    const prev = Atomics.compareExchange(words, CursorStateIndex.GENERATION, start | 0, desired | 0) >>> 0;
    if (prev === start) break;
    start = prev;
  }

  // Store the payload fields.
  Atomics.store(words, CursorStateIndex.ENABLE, update.enable | 0);
  Atomics.store(words, CursorStateIndex.X, update.x | 0);
  Atomics.store(words, CursorStateIndex.Y, update.y | 0);
  Atomics.store(words, CursorStateIndex.HOT_X, update.hotX | 0);
  Atomics.store(words, CursorStateIndex.HOT_Y, update.hotY | 0);
  Atomics.store(words, CursorStateIndex.WIDTH, update.width | 0);
  Atomics.store(words, CursorStateIndex.HEIGHT, update.height | 0);
  Atomics.store(words, CursorStateIndex.PITCH_BYTES, update.pitchBytes | 0);
  Atomics.store(words, CursorStateIndex.FORMAT, update.format | 0);
  Atomics.store(words, CursorStateIndex.BASE_PADDR_LO, update.basePaddrLo | 0);
  Atomics.store(words, CursorStateIndex.BASE_PADDR_HI, update.basePaddrHi | 0);

  // Final publish step: increment generation and clear the busy bit.
  const newGeneration = (((start + 1) >>> 0) & (~CURSOR_STATE_GENERATION_BUSY_BIT >>> 0)) >>> 0;
  Atomics.store(words, CursorStateIndex.GENERATION, newGeneration | 0);
  return newGeneration;
}

File: crates/aero-shared/src/cursor_state.rs
//! Lock-free hardware cursor descriptor shared between multiple workers.
//!
//! This structure is designed to be shared with JavaScript via
//! `SharedArrayBuffer` + `Int32Array`, using atomic operations only.
//!
//! ## Publish protocol
//!
//! This uses the same seqlock-style scheme as [`crate::scanout_state::ScanoutState`]:
//! the high bit of `generation` is treated as a "busy" marker:
//! - The writer sets [`CURSOR_STATE_GENERATION_BUSY_BIT`] before writing fields.
//! - The writer stores the new committed generation (busy bit cleared) as the last step.
//! - Readers spin/retry if the busy bit is set or if the generation changes mid-snapshot.

#[cfg(all(feature = "loom", test))]
use loom::sync::atomic::AtomicU32;
#[cfg(not(all(feature = "loom", test)))]
use std::sync::atomic::AtomicU32;

use std::sync::atomic::Ordering;

use aero_protocol::aerogpu::aerogpu_pci::AerogpuFormat;

/// Cursor format values use the AeroGPU `AerogpuFormat` (`u32`) discriminants.
///
/// Semantics (from the AeroGPU protocol):
/// - `*X8*` formats (`B8G8R8X8*`, `R8G8B8X8*`) do not carry alpha. When converting
///   to RGBA (e.g. for cursor blending), treat alpha as fully opaque (`0xFF`) and
///   ignore the stored `X` byte.
/// - `*_SRGB` variants are layout-identical to their UNORM counterparts; only
///   the color space interpretation differs. Presenters must avoid
///   double-applying gamma when handling sRGB cursor formats.
///
/// This must stay in sync with `aero_protocol::aerogpu::aerogpu_pci::AerogpuFormat`.
pub const CURSOR_FORMAT_B8G8R8A8: u32 = AerogpuFormat::B8G8R8A8Unorm as u32;
pub const CURSOR_FORMAT_B8G8R8X8: u32 = AerogpuFormat::B8G8R8X8Unorm as u32;
pub const CURSOR_FORMAT_R8G8B8A8: u32 = AerogpuFormat::R8G8B8A8Unorm as u32;
pub const CURSOR_FORMAT_R8G8B8X8: u32 = AerogpuFormat::R8G8B8X8Unorm as u32;

/// Internal bit used to mark `generation` as "being updated".
///
/// The published generation value (the one returned from [`CursorState::snapshot`]) never has
/// this bit set and increments by 1 per completed update.
pub const CURSOR_STATE_GENERATION_BUSY_BIT: u32 = 1 << 31;

/// The cursor state is an array of 32-bit words to keep it trivially shareable
/// with JS as an `Int32Array`.
pub const CURSOR_STATE_U32_LEN: usize = 12;
pub const CURSOR_STATE_BYTE_LEN: usize = CURSOR_STATE_U32_LEN * 4;

pub mod header_index {
    //! Indices into the cursor state when viewed as a `u32[]` / `Int32Array`.

    pub const GENERATION: usize = 0;
    pub const ENABLE: usize = 1;
    pub const X: usize = 2;
    pub const Y: usize = 3;
    pub const HOT_X: usize = 4;
    pub const HOT_Y: usize = 5;
    pub const WIDTH: usize = 6;
    pub const HEIGHT: usize = 7;
    pub const PITCH_BYTES: usize = 8;
    pub const FORMAT: usize = 9;
    pub const BASE_PADDR_LO: usize = 10;
    pub const BASE_PADDR_HI: usize = 11;
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct CursorStateUpdate {
    pub enable: u32,
    /// Cursor position in the scanout coordinate space (top-left origin).
    ///
    /// Stored as an i32 bit-pattern in a u32 word.
    pub x: i32,
    /// Cursor position in the scanout coordinate space (top-left origin).
    ///
    /// Stored as an i32 bit-pattern in a u32 word.
    pub y: i32,
    pub hot_x: u32,
    pub hot_y: u32,
    pub width: u32,
    pub height: u32,
    pub pitch_bytes: u32,
    pub format: u32,
    pub base_paddr_lo: u32,
    pub base_paddr_hi: u32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct CursorStateSnapshot {
    pub generation: u32,
    pub enable: u32,
    pub x: i32,
    pub y: i32,
    pub hot_x: u32,
    pub hot_y: u32,
    pub width: u32,
    pub height: u32,
    pub pitch_bytes: u32,
    pub format: u32,
    pub base_paddr_lo: u32,
    pub base_paddr_hi: u32,
}

impl CursorStateSnapshot {
    pub fn base_paddr(self) -> u64 {
        (self.base_paddr_hi as u64) << 32 | self.base_paddr_lo as u64
    }
}

#[repr(C)]
pub struct CursorState {
    /// Sequence counter used to publish updates.
    ///
    /// The high bit ([`CURSOR_STATE_GENERATION_BUSY_BIT`]) is used internally to mark an
    /// in-progress update; published generations always have the bit cleared.
    pub generation: AtomicU32,

    pub enable: AtomicU32,
    pub x: AtomicU32,
    pub y: AtomicU32,
    pub hot_x: AtomicU32,
    pub hot_y: AtomicU32,
    pub width: AtomicU32,
    pub height: AtomicU32,
    pub pitch_bytes: AtomicU32,
    pub format: AtomicU32,
    pub base_paddr_lo: AtomicU32,
    pub base_paddr_hi: AtomicU32,
}

impl CursorState {
    pub fn new() -> Self {
        Self {
            generation: AtomicU32::new(0),
            enable: AtomicU32::new(0),
            x: AtomicU32::new(0),
            y: AtomicU32::new(0),
            hot_x: AtomicU32::new(0),
            hot_y: AtomicU32::new(0),
            width: AtomicU32::new(0),
            height: AtomicU32::new(0),
            pitch_bytes: AtomicU32::new(0),
            format: AtomicU32::new(CURSOR_FORMAT_B8G8R8A8),
            base_paddr_lo: AtomicU32::new(0),
            base_paddr_hi: AtomicU32::new(0),
        }
    }

    /// Publish a complete cursor update.
    ///
    /// Protocol:
    /// 1) Mark the state as "in progress" by setting [`CURSOR_STATE_GENERATION_BUSY_BIT`].
    /// 2) Store all non-generation fields.
    /// 3) Increment `generation` (busy bit cleared) as the final publish step.
    pub fn publish(&self, update: CursorStateUpdate) -> u32 {
        // Acquire the write lock by setting the busy bit.
        let mut start = self.generation.load(Ordering::SeqCst);
        loop {
            if start & CURSOR_STATE_GENERATION_BUSY_BIT != 0 {
                std::hint::spin_loop();
                start = self.generation.load(Ordering::SeqCst);
                continue;
            }

            match self.generation.compare_exchange_weak(
                start,
                start | CURSOR_STATE_GENERATION_BUSY_BIT,
                Ordering::SeqCst,
                Ordering::SeqCst,
            ) {
                Ok(_) => break,
                Err(actual) => start = actual,
            }
        }

        test_yield();

        self.enable.store(update.enable, Ordering::SeqCst);
        test_yield();
        self.x.store(update.x as u32, Ordering::SeqCst);
        test_yield();
        self.y.store(update.y as u32, Ordering::SeqCst);
        test_yield();
        self.hot_x.store(update.hot_x, Ordering::SeqCst);
        test_yield();
        self.hot_y.store(update.hot_y, Ordering::SeqCst);
        test_yield();
        self.width.store(update.width, Ordering::SeqCst);
        test_yield();
        self.height.store(update.height, Ordering::SeqCst);
        test_yield();
        self.pitch_bytes.store(update.pitch_bytes, Ordering::SeqCst);
        test_yield();
        self.format.store(update.format, Ordering::SeqCst);
        test_yield();
        self.base_paddr_lo
            .store(update.base_paddr_lo, Ordering::SeqCst);
        test_yield();
        self.base_paddr_hi
            .store(update.base_paddr_hi, Ordering::SeqCst);

        test_yield();

        // Final publish step: increment generation and clear the busy bit.
        let new_generation = start.wrapping_add(1) & !CURSOR_STATE_GENERATION_BUSY_BIT;
        self.generation.store(new_generation, Ordering::SeqCst);
        new_generation
    }

    pub fn snapshot(&self) -> CursorStateSnapshot {
        loop {
            let gen0 = self.generation.load(Ordering::SeqCst);
            if gen0 & CURSOR_STATE_GENERATION_BUSY_BIT != 0 {
                // Writer in progress.
                std::hint::spin_loop();
                test_yield();
                continue;
            }

            let enable = self.enable.load(Ordering::SeqCst);
            let x = self.x.load(Ordering::SeqCst) as i32;
            let y = self.y.load(Ordering::SeqCst) as i32;
            let hot_x = self.hot_x.load(Ordering::SeqCst);
            let hot_y = self.hot_y.load(Ordering::SeqCst);
            let width = self.width.load(Ordering::SeqCst);
            let height = self.height.load(Ordering::SeqCst);
            let pitch_bytes = self.pitch_bytes.load(Ordering::SeqCst);
            let format = self.format.load(Ordering::SeqCst);
            let base_paddr_lo = self.base_paddr_lo.load(Ordering::SeqCst);
            let base_paddr_hi = self.base_paddr_hi.load(Ordering::SeqCst);

            let gen1 = self.generation.load(Ordering::SeqCst);
            if gen0 != gen1 {
                test_yield();
                continue;
            }

            return CursorStateSnapshot {
                generation: gen0,
                enable,
                x,
                y,
                hot_x,
                hot_y,
                width,
                height,
                pitch_bytes,
                format,
                base_paddr_lo,
                base_paddr_hi,
            };
        }
    }
}

impl Default for CursorState {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(all(test, feature = "loom"))]
#[inline]
fn test_yield() {
    loom::thread::yield_now();
}

#[cfg(all(test, not(feature = "loom")))]
#[inline]
fn test_yield() {
    std::thread::yield_now();
}

#[cfg(not(test))]
#[inline]
fn test_yield() {}

#[cfg(all(test, not(feature = "loom")))]
mod tests {
    use super::*;

    use std::sync::atomic::AtomicBool;
    use std::sync::Arc;
    use std::thread;

    #[test]
    fn cursor_format_constants_match_aerogpu_format_discriminants() {
        assert_eq!(CURSOR_FORMAT_B8G8R8A8, AerogpuFormat::B8G8R8A8Unorm as u32);
        assert_eq!(CURSOR_FORMAT_B8G8R8X8, AerogpuFormat::B8G8R8X8Unorm as u32);
        assert_eq!(CURSOR_FORMAT_R8G8B8A8, AerogpuFormat::R8G8B8A8Unorm as u32);
        assert_eq!(CURSOR_FORMAT_R8G8B8X8, AerogpuFormat::R8G8B8X8Unorm as u32);
    }

    #[test]
    fn cursor_state_struct_matches_declared_u32_len() {
        assert_eq!(core::mem::size_of::<CursorState>(), CURSOR_STATE_BYTE_LEN);
    }

    #[test]
    fn header_indices_match_struct_layout() {
        let state = CursorState::new();
        let base = core::ptr::addr_of!(state) as usize;

        let field_offset = |ptr: *const AtomicU32| ptr as usize - base;

        assert_eq!(
            field_offset(core::ptr::addr_of!(state.generation)),
            header_index::GENERATION * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.enable)),
            header_index::ENABLE * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.x)),
            header_index::X * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.y)),
            header_index::Y * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.hot_x)),
            header_index::HOT_X * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.hot_y)),
            header_index::HOT_Y * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.width)),
            header_index::WIDTH * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.height)),
            header_index::HEIGHT * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.pitch_bytes)),
            header_index::PITCH_BYTES * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.format)),
            header_index::FORMAT * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.base_paddr_lo)),
            header_index::BASE_PADDR_LO * 4
        );
        assert_eq!(
            field_offset(core::ptr::addr_of!(state.base_paddr_hi)),
            header_index::BASE_PADDR_HI * 4
        );
    }

    #[test]
    fn generation_increments_by_one_per_completed_publish() {
        let state = CursorState::new();

        let g0 = state.snapshot().generation;
        state.publish(CursorStateUpdate {
            enable: 1,
            x: 1,
            y: 2,
            hot_x: 3,
            hot_y: 4,
            width: 5,
            height: 6,
            pitch_bytes: 7,
            format: CURSOR_FORMAT_B8G8R8A8,
            base_paddr_lo: 8,
            base_paddr_hi: 9,
        });
        let g1 = state.snapshot().generation;
        state.publish(CursorStateUpdate {
            enable: 0,
            x: 10,
            y: 11,
            hot_x: 12,
            hot_y: 13,
            width: 14,
            height: 15,
            pitch_bytes: 16,
            format: CURSOR_FORMAT_B8G8R8X8,
            base_paddr_lo: 17,
            base_paddr_hi: 18,
        });
        let g2 = state.snapshot().generation;

        assert_eq!(g1, g0.wrapping_add(1));
        assert_eq!(g2, g1.wrapping_add(1));
    }

    #[test]
    fn generation_wraps_without_leaking_busy_bit() {
        let state = CursorState::new();

        // Force the generation near the busy-bit boundary (high bit is reserved).
        state.generation.store(0x7fff_fffe, Ordering::SeqCst);

        let g0 = state.snapshot().generation;
        assert_eq!(g0, 0x7fff_fffe);

        let g1 = state.publish(CursorStateUpdate {
            enable: 0,
            x: 0,
            y: 0,
            hot_x: 0,
            hot_y: 0,
            width: 0,
            height: 0,
            pitch_bytes: 0,
            format: CURSOR_FORMAT_B8G8R8A8,
            base_paddr_lo: 0,
            base_paddr_hi: 0,
        });
        assert_eq!(g1, 0x7fff_ffff);
        assert_eq!(g1 & CURSOR_STATE_GENERATION_BUSY_BIT, 0);

        let g2 = state.publish(CursorStateUpdate {
            enable: 0,
            x: 0,
            y: 0,
            hot_x: 0,
            hot_y: 0,
            width: 0,
            height: 0,
            pitch_bytes: 0,
            format: CURSOR_FORMAT_B8G8R8A8,
            base_paddr_lo: 0,
            base_paddr_hi: 0,
        });
        assert_eq!(g2, 0x0000_0000);
        assert_eq!(g2 & CURSOR_STATE_GENERATION_BUSY_BIT, 0);
    }

    #[test]
    fn snapshot_is_coherent_across_concurrent_updates() {
        let state = Arc::new(CursorState::new());

        // Initialize to a non-zero state so the reader doesn't have to special-case the default.
        state.publish(CursorStateUpdate {
            enable: 1,
            x: 0,
            y: 1,
            hot_x: 2,
            hot_y: 3,
            width: 4,
            height: 5,
            pitch_bytes: 6,
            format: CURSOR_FORMAT_B8G8R8A8,
            base_paddr_lo: 7,
            base_paddr_hi: 8,
        });

        let start = Arc::new(std::sync::Barrier::new(2));
        let done = Arc::new(AtomicBool::new(false));

        let writer_state = state.clone();
        let writer_start = start.clone();
        let writer_done = done.clone();
        let writer = thread::spawn(move || {
            writer_start.wait();

            for token in 0u32..10_000 {
                writer_state.publish(CursorStateUpdate {
                    enable: 1,
                    x: token as i32,
                    y: token.wrapping_add(1) as i32,
                    hot_x: token.wrapping_add(2),
                    hot_y: token.wrapping_add(3),
                    width: token.wrapping_add(4),
                    height: token.wrapping_add(5),
                    pitch_bytes: token.wrapping_add(6),
                    format: CURSOR_FORMAT_B8G8R8A8,
                    base_paddr_lo: token.wrapping_add(7),
                    base_paddr_hi: token.wrapping_add(8),
                });
            }

            writer_done.store(true, Ordering::SeqCst);
        });

        let reader_state = state.clone();
        let reader_start = start.clone();
        let reader_done = done.clone();
        let reader = thread::spawn(move || {
            reader_start.wait();

            while !reader_done.load(Ordering::SeqCst) {
                let snap = reader_state.snapshot();

                assert_eq!(snap.format, CURSOR_FORMAT_B8G8R8A8);
                assert_eq!(snap.enable, 1);

                let token = snap.width.wrapping_sub(4);
                assert_eq!(snap.x, token as i32);
                assert_eq!(snap.y, token.wrapping_add(1) as i32);
                assert_eq!(snap.hot_x, token.wrapping_add(2));
                assert_eq!(snap.hot_y, token.wrapping_add(3));
                assert_eq!(snap.width, token.wrapping_add(4));
                assert_eq!(snap.height, token.wrapping_add(5));
                assert_eq!(snap.pitch_bytes, token.wrapping_add(6));
                assert_eq!(snap.base_paddr_lo, token.wrapping_add(7));
                assert_eq!(snap.base_paddr_hi, token.wrapping_add(8));
            }

            // One last snapshot after the writer has finished.
            let snap = reader_state.snapshot();
            assert_eq!(snap.format, CURSOR_FORMAT_B8G8R8A8);
            assert_eq!(snap.enable, 1);
        });

        writer.join().unwrap();
        reader.join().unwrap();
    }
}

#[cfg(all(test, feature = "loom"))]
mod loom_tests {
    use super::*;

    use loom::sync::Arc;
    use loom::thread;

    #[test]
    fn snapshot_never_observes_partial_publish() {
        loom::model(|| {
            let state = Arc::new(CursorState::new());

            // Initialize to a coherent state using the same invariant relation as the update.
            state.publish(CursorStateUpdate {
                enable: 1,
                x: 0,
                y: 1,
                hot_x: 2,
                hot_y: 3,
                width: 4,
                height: 5,
                pitch_bytes: 6,
                format: CURSOR_FORMAT_B8G8R8A8,
                base_paddr_lo: 7,
                base_paddr_hi: 8,
            });

            let writer_state = state.clone();
            let reader_state = state.clone();

            let writer = thread::spawn(move || {
                writer_state.publish(CursorStateUpdate {
                    enable: 1,
                    x: 1,
                    y: 2,
                    hot_x: 3,
                    hot_y: 4,
                    width: 5,
                    height: 6,
                    pitch_bytes: 7,
                    format: CURSOR_FORMAT_B8G8R8A8,
                    base_paddr_lo: 8,
                    base_paddr_hi: 9,
                });
            });

            let reader = thread::spawn(move || {
                let snap = reader_state.snapshot();

                assert_eq!(snap.format, CURSOR_FORMAT_B8G8R8A8);
                assert_eq!(snap.enable, 1);

                let token = snap.width.wrapping_sub(4);
                assert_eq!(snap.x, token as i32);
                assert_eq!(snap.y, token.wrapping_add(1) as i32);
                assert_eq!(snap.hot_x, token.wrapping_add(2));
                assert_eq!(snap.hot_y, token.wrapping_add(3));
                assert_eq!(snap.width, token.wrapping_add(4));
                assert_eq!(snap.height, token.wrapping_add(5));
                assert_eq!(snap.pitch_bytes, token.wrapping_add(6));
                assert_eq!(snap.base_paddr_lo, token.wrapping_add(7));
                assert_eq!(snap.base_paddr_hi, token.wrapping_add(8));
            });

            writer.join().unwrap();
            reader.join().unwrap();
        });
    }
}
